### PR TITLE
feat: copilot controller

### PR DESCRIPTION
**Pull request by darkskygit**
fix CLOUD-27

### CODE REVIEW

**Comment by darkskygit** [comment]
* **#6528** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6528?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6527** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6527?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6506** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6506?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6525** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6525?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6445** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6445?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6448** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6448?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6437** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6437?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6272** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6272?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6230** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6254** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6254?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6241** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6241?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6229** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6229?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @darkskygit and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/backend/server/src/core/quota/schema.ts [hunk 1]
```diff
@@ -93,11 +93,35 @@ export const Quotas: Quota[] = [
...
+      // NOTE: this is a product decision, may change in future
+      businessBlobLimit: 100 * OneMB,
+      // total blob limit 10GB
+      storageQuota: 10 * OneGB,
+      // history period of validity 7 days
+      historyPeriod: 7 * OneDay,
+      // member limit 3
+      memberLimit: 3,
+      // copilot action limit 10
+      copilotActionLimit: 10,
```
Comment (forehalo) [inline packages/backend/server/src/core/quota/schema.ts]: it should be 30 if I remember correctly?

File: packages/backend/server/src/fundamentals/graphql/index.ts [hunk 2]
```diff
@@ -23,6 +23,15 @@ export type GraphqlContext = {
     GraphQLModule.forRootAsync<ApolloDriverConfig>({
       driver: ApolloDriver,
       useFactory: (config: Config) => {
+        const copilotAuthorization = config.featureFlags.copilotAuthorization;
+        const cors = {
+          cors: {
+            origin: [
+              'https://try-blocksuite.vercel.app/',
```
Comment (forehalo) [inline packages/backend/server/src/fundamentals/graphql/index.ts]: remember to remove before merging

File: packages/backend/server/src/plugins/copilot/session.ts [hunk 3]
```diff
@@ -170,16 +211,138 @@ export class ChatSessionService {
...
+      return total + encoder.encode_ordinary(m.content).length;
+    }, 0);
+  }
+
+  async countSessions(
+    userId: string,
+    workspaceId: string,
+    options?: { docId?: string; action?: boolean }
+  ): Promise<number> {
+    // NOTE: only used for anonymous session in development
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/session.ts]: remember to remove before merging

File: packages/backend/server/src/plugins/copilot/session.ts [hunk 4]
```diff
@@ -170,16 +211,138 @@ export class ChatSessionService {
...
+    });
+  }
+
+  async listSessions(
+    userId: string | undefined,
+    workspaceId: string,
+    options?: { docId?: string; action?: boolean }
+  ): Promise<string[]> {
+    // NOTE: only used for anonymous session in development
+    if (!userId && AFFiNE.featureFlags.copilotAuthorization) {
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/session.ts]: remember to remove

File: packages/backend/server/src/plugins/copilot/session.ts [hunk 5]
```diff
@@ -170,16 +211,138 @@ export class ChatSessionService {
...
+  }
+
+  async listHistories(
+    userId: string | undefined,
+    workspaceId: string,
+    docId?: string,
+    options?: ListHistoriesOptions
+  ): Promise<ChatHistory[]> {
+    // NOTE: only used for anonymous session in development
+    if (!userId && AFFiNE.featureFlags.copilotAuthorization) {
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/session.ts]: remove

File: packages/backend/server/src/core/quota/schema.ts (hunk 1 above)
Comment (darkskygit) [reply packages/backend/server/src/core/quota/schema.ts]: see https://whimsical.com/affine-ai-Jt3idZd6njecW3ZLFwHjFe

File: packages/backend/server/src/plugins/copilot/session.ts (hunk 4 above)
Comment (forehalo) [reply packages/backend/server/src/plugins/copilot/session.ts]: could you move all authentication ignore code to a separated PR? we will easily delete them all by reverting it.

Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/session.ts]: lgtm

Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/session.ts]: moved to https://github.com/toeverything/AFFiNE/pull/6445

File: packages/backend/server/src/plugins/copilot/resolver.ts [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+@InputType()
+class CreateChatSessionInput {
+  @Field(() => String)
+  workspaceId!: string;
+
+  @Field(() => String)
+  docId!: string;
+
+  @Field(() => Boolean, {
```
Comment (pengx17) [inline packages/backend/server/src/plugins/copilot/resolver.ts]: This should be `() => String`

File: .github/actions/deploy/deploy.mjs [hunk 7]
```diff
@@ -116,7 +116,7 @@ const createHelmCommand = ({ isDryRun }) => {
     `--set        graphql.app.experimental.enableJwstCodec=${namespace === 'dev'}`,
     `--set        graphql.app.features.earlyAccessPreview=false`,
     `--set        graphql.app.features.syncClientVersionCheck=true`,
-    `--set        graphql.app.features.copilotAuthorization=true`,
+    `--set        graphql.app.features.copilotAuthorization=${namespace === 'dev'}`,
```
Comment (forehalo) [inline .github/actions/deploy/deploy.mjs]: move

File: packages/backend/server/src/plugins/copilot/resolver.ts [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    private readonly chatSession: ChatSessionService
+  ) {}
+
+  @ResolveField(() => CopilotQuotaType, {
+    name: 'quota',
+    description: 'Get the quota of the user in the workspace',
+    complexity: 2,
+  })
+  async getQuota(
+    @Parent() _copilot: CopilotType,
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/resolver.ts]: not required

**Review by forehalo — APPROVED** [review]

**Pull request merged**